lib/recommendation.js

```diff
@@ -0,0 +1,61 @@
+var XMLHttpRequest = Cc['@mozilla.org/xmlextras/xmlhttprequest;1'].createInstance();
+
+
+class RecommendationServer {
+  /*
+  Class providing a connection to a recommendation server and a method for
+  querying it. This class should be initialized once and exported as a
+  singleton to modules that will use it.
+
+  Usage:
+
+  let recommendation = new RecommendationServer('http://universal-search.dev');
+
+  */
+  constructor(server) {
+    this.server = server;
+  }
+
+  _url(query) {
+    return `${this.server}/?q=${encodeURIComponent(query)}`;
+  }
+
+  make(query) {
+    /*
+    Method to query a recommendation server. Returns a promise that resolves to
+    a recommendation or rejects to a reason that a recommendation is not
+    available.
+
+    Usage:
+
+    recommendation.make('hello wor').then(data => {
+      // Recommendation returned as `data` object literal.
+    }, reason => {
+      // No recommendation returned for `reason` string.
+    });
+
+    */
+    const url = this._url(query);
+    return new Promise((resolve, reject) => {
+      try {
+        var req = new XMLHttpRequest();
+        req.addEventListener('abort', (evt) => {
+          reject('Aborted request.');
+        });
+        req.addEventListener('error', (evt) => {
+          reject('Errored request.');
+        });
+        req.addEventListener('load', (evt) => {
+          if (req.status === 200 && req.responseText !== {}) {
+            resolve(JSON.parse(req.responseText));
+          }
+          reject('No recommendation available.');
+        });
+        req.open('GET', url);
+        req.send();
```

Comment:
Might be good to set a timeout before sending the request? That could also be specified in the constructor.